python/cudf/cudf/core/df_protocol.py

```diff
@@ -0,0 +1,825 @@
+_CP_DTYPES = {0: _INTS, 1: _UINTS, 2: _FLOATS, 20: {8: bool}}
+
+
+def from_dataframe(
+    df: DataFrameObject, allow_copy: bool = False
+) -> _CuDFDataFrame:
+    """
+    Construct a cudf DataFrame from ``df`` if it supports ``__dataframe__``
+    """
+    if isinstance(df, cudf.DataFrame):
+        return df
+
+    if not hasattr(df, "__dataframe__"):
+        raise ValueError("`df` does not support __dataframe__")
+
+    return _from_dataframe(df.__dataframe__(allow_copy=allow_copy))
+
+
+def _from_dataframe(df: DataFrameObject) -> _CuDFDataFrame:
+    """
+    Create a cudf DataFrame object from DataFrameObject.
+    """
+    # Check number of chunks, if there's more than one we need to iterate
+    if df.num_chunks() > 1:
+        raise NotImplementedError("More than one chunk not handled yet")
+
+    # We need a dict of columns here, with each column being a cudf column.
+    columns = dict()
+    _k = _DtypeKind
+    _buffers = []  # hold on to buffers, keeps memory alive
+    for name in df.column_names():
+        col = df.get_column_by_name(name)
+
+        if col.dtype[0] in (_k.INT, _k.UINT, _k.FLOAT, _k.BOOL):
+            columns[name], _buf = _protocol_to_cudf_column_numeric(col)
+
+        elif col.dtype[0] == _k.CATEGORICAL:
+            columns[name], _buf = _protocol_to_cudf_column_categorical(col)
+
+        elif col.dtype[0] == _k.STRING:
+            columns[name], _buf = _protocol_to_cudf_column_string(col)
+
+        else:
+            raise NotImplementedError(
+                f"Data type {col.dtype[0]} not handled yet"
+            )
+
+        _buffers.append(_buf)
+
+    df_new = cudf.DataFrame._from_data(columns)
+    df_new._buffers = _buffers
+    return df_new
+
+
+def _protocol_to_cudf_column_numeric(
+    col: _CuDFColumn,
+) -> Tuple[
+    cudf.core.column.ColumnBase,
+    Mapping[str, Optional[Tuple[_CuDFBuffer, ProtoDtype]]],
+]:
+    """
+    Convert an int, uint, float or bool protocol column
+    to the corresponding cudf column
+    """
+    if col.offset != 0:
+        raise NotImplementedError("column.offset > 0 not handled yet")
+
+    buffers = col.get_buffers()
+    assert buffers["data"] is not None, "data buffer should not be None"
+    _dbuffer, _ddtype = buffers["data"]
+    _check_buffer_is_on_gpu(_dbuffer)
+    cudfcol_num = build_column(
+        Buffer(_dbuffer.ptr, _dbuffer.bufsize),
+        protocol_dtype_to_cupy_dtype(_ddtype),
+    )
+    return _set_missing_values(col, cudfcol_num), buffers
+
+
+def _check_buffer_is_on_gpu(buffer: _CuDFBuffer) -> None:
+    if (
+        buffer.__dlpack_device__()[0] != _Device.CUDA
+        and not buffer._allow_copy
+    ):
+        raise TypeError(
+            "This operation must copy data from CPU to GPU. "
+            "Set `allow_copy=True` to allow it."
+        )
+
+    elif buffer.__dlpack_device__()[0] != _Device.CUDA and buffer._allow_copy:
+        raise NotImplementedError(
+            "Only cuDF/GPU dataframes are supported for now."
```

Comment:
Multi-line string needs a space between sentences.
```suggestion
            "Only cuDF/GPU dataframes are supported for now. "
```